crates/modelardb_compression_python/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+
+mod convert;
+
+use std::sync::Arc;
+
+use arrow::array::{Array, ArrayRef, BinaryArray, Float32Array, UInt64Array, UInt8Array};
+use arrow::datatypes::Schema;
+use arrow::error::ArrowError;
+use arrow::record_batch::RecordBatch;
+use modelardb_common::schemas::{COMPRESSED_SCHEMA, QUERY_SCHEMA, UNCOMPRESSED_SCHEMA};
+use modelardb_common::types::{TimestampArray, TimestampBuilder, ValueArray, ValueBuilder};
+use modelardb_compression::{self, models};
+use pyo3::import_exception;
+use pyo3::prelude::{pyfunction, pymodule, wrap_pyfunction, PyAny, PyModule, PyResult, Python};
+
+import_exception!(pyarrow, ArrowException);
+
+/// Initialize the module by registering functions in the Python module.
+#[pymodule]
+fn modelardb_compression_python(_python: Python<'_>, py_module: &PyModule) -> PyResult<()> {
+    py_module.add_function(wrap_pyfunction!(compress, py_module)?)?;
+    py_module.add_function(wrap_pyfunction!(decompress, py_module)?)?;
+
+    Ok(())
+}
+
+/// Compress a [`RecordBatch`] containing an univariate time series within `error_bound`. An
+/// `univariate_id` can optionally be set to specify which univariate time series the compressed
+/// segments was produced from, if none is set zero is used. An [`ArrowException`] is returned if
+/// the schema of `uncompressed` does not match [`UNCOMPRESSED_SCHEMA`] or if [`pyo3`] cannot
+/// convert the arguments from Python to Rust and the result from Rust to Python.
+#[pyfunction(univariate_id = 0)]
+fn compress<'a>(
+    uncompressed: &'a PyAny,
+    error_bound: &'a PyAny,
+    univariate_id: u64,
+    python: Python<'a>,
+) -> PyResult<&'a PyAny> {
+    // Convert the record batch from Python to Rust and validate the arguments.
+    let uncompressed = convert::python_record_batch_to_rust(uncompressed)?;
+    if !fields_has_equal_types(&uncompressed.schema(), &UNCOMPRESSED_SCHEMA.0) {
+        return Err(ArrowException::new_err(
+            "Uncompressed can only contain millisecond timestamps and float32 values.".to_owned(),
+        ));
+    }
+
+    let error_bound = models::ErrorBound::try_new(error_bound.extract::<f32>()?)
+        .map_err(|error| convert::to_py_err(ArrowError::InvalidArgumentError(error.to_string())))?;
+
+    // Compress the timestamps and values to segments containing metadata and models.
+    let timestamps = modelardb_common::array!(uncompressed, 0, TimestampArray);
+    let values = modelardb_common::array!(uncompressed, 1, ValueArray);
+
+    // unwrap() is safe as the timestamps and values are guaranteed to be the same length.
+    let compressed = modelardb_compression::try_compress(
+        univariate_id,
+        timestamps,
+        values,
+        error_bound,
+        &COMPRESSED_SCHEMA,
+    )
+    .unwrap();
+
+    // Convert and return the finished compressed PyArrow record batch.
+    convert::rust_record_batch_to_python(compressed, python)
+}
+
+/// Decompress a [`RecordBatch`] containing compressed segments to one more univariate time series.
```

Comment:
Should this be `to one or more` instead?

Reply:
Fixed.